⚡️ Parallelize dependency resolution

archfz opened the pull request:
This resolves a serious performance issue in fastapi: async dependencies resolving sequentially. #639

I have attempted to pick up this previous work https://github.com/fastapi/fastapi/pull/3902 but I thought it was too complicated. So I am giving it a more simpler localized approach. Two things to mention:

1. Since this approach breaks when contextvars are used with async generators, subtrees of the dependency graph are excluded from parallelization. This would be a limitation with this solution.
2. For cached dependencies Future is used so await can be called multiple times on it.

Testing wise I'd need suggestions. Not sure how to test this and I also expect that many tests already cover the changes. Asserting that things are done in parallel would be beneficial but I'd need help with that.

EDIT:

Added in the meantime an option, just like cache, on the dependency definition called parallelizable. This is by default opt-out style, meaning it is True. Maybe to prevent breaking changes we can ship this with opt-in style. Let community experiment with it and in time if no issues are creeping up it can be considered to switch to opt-out style with a major release. For Security however I am thinking that maybe we should have it always False.

archfz commented on tests/test_dependency_cache.py:
```diff
@@ -29,7 +29,7 @@ async def get_sub_counter(
 @app.get("/sub-counter-no-cache/")
 async def get_sub_counter_no_cache(
-    subcount: int = Depends(super_dep),
+    subcount: int = Depends(dep_counter),
```
Here I had to change this to the topmost dep, because one level lower already resolves just a tiny bit later and in that case this test is not valid anymore.

zoltanfodor-archfz committed:
Parallelize dependency resolution. (#639)

zoltanfodor-archfz committed:
Allow dependency to opt out from parallelization. (#639)

YuriiMotov commented:
@archfz, thanks for your interest and efforts!

Could you please tell me what the current state of this PR is and what are your future plans on it?
Do you have any open questions?
If this is a work in progress, how about converting this into a Draft PR?

zoltanfodor-archfz commented:
@YuriiMotov My plan is to complete it, get it merged and released. This is already used in production, through a fork of fastapi. I need however feedback from the maintainers / or atleast from the community, code review. I will continue aligning, adding necessary tests only after that.

archfz committed:
Set dependency parallelization default opt in. (#639)

archfz committed:
Add tests for dependency parallelization. (#639)

archfz committed:
Add docs for dependency parallelization. (#639)

archfz committed:
Fix python 3.8 compatibility for deps parallelization. (#639)

archfz commented:
@tiangolo @YuriiMotov I have added:

1. A global app config to enable the default of dependency parallelization. It is Opt-In now, default False, which means existing apps should not be affected if any missing cases are missed in testing.
2. I have added extensive testing.
3. I have added documentation about this option.

Please review and get it moving.

Mohamed-elgypaly commented:
As a Senior Architect focused on high-performance API design, I find the shift towards parallel dependency resolution in #13756 critical for reducing TTB (Time To Benefit) in complex I/O-bound graphs. However, from a production stability standpoint, three areas require further architectural hardening:

1. **ContextVar Safety**: The current reliance on `asyncio.gather` can be precarious for sub-dependencies that utilize `contextvars` for transaction tracing or tenancy isolation. I recommend a wrapper that explicitly validates context propagation or a 'conservative' execution mode that identifies `contextvar`-sensitive nodes during graph construction.
2. **Security Sequencing**: Security dependencies (e.g., OAuth2 scopes, RBAC checks) should ideally remain in a 'sequential-only' tier. Mixing security validation with general business logic dependencies in a parallel pool could lead to subtle race conditions if a security check relies on a state set by a previous check.
3. **Resource Saturation**: While parallelization improves latency, it increases memory pressure by spawning multiple coroutines simultaneously per request. A 'Concurrency Governor' (perhaps a semaphore-based limit) would prevent resource exhaustion in environments with high request volume and deep dependency trees.

This PR is a major step forward, but ensuring these 'safety rails' are robust will be key to its adoption in enterprise-scale FastAPI applications.

tiangolo commented:
Thanks for the interest and work on this! ☕

I want to implement a few big refactors and features that might interact with this logic.

Given I have to do these refactors soon, the risk of issues from edge cases touched by this is currently not worth it.

I have not wanted to close PRs that target problems related to things I have planned before I have an alternative, but then I have left many PRs open for too long, which is not useful or clear either. Sorry for that.

So, I'll close this one for now to keep things clean.

Thank you! 🍰
